concordium-consensus/src/Concordium/KonsensusV1/TreeState/StartUp.hs

```diff
@@ -0,0 +1,208 @@
+{-# LANGUAGE TypeFamilies #-}
+
+module Concordium.KonsensusV1.TreeState.StartUp where
+
+import Control.Monad.Catch
+import Control.Monad.IO.Class
+import Lens.Micro.Platform
+
+import Concordium.Types
+import Concordium.Types.HashableTo
+import Concordium.Types.Parameters hiding (getChainParameters)
+
+import Concordium.GlobalState.BlockState
+import Concordium.GlobalState.Parameters hiding (getChainParameters)
+import qualified Concordium.GlobalState.Persistent.BlockState as PBS
+import qualified Concordium.GlobalState.Statistics as Stats
+import qualified Concordium.GlobalState.TransactionTable as TT
+import qualified Concordium.GlobalState.Types as GSTypes
+import Concordium.KonsensusV1.Consensus
+import Concordium.KonsensusV1.TreeState.Implementation
+import qualified Concordium.KonsensusV1.TreeState.LowLevel as LowLevel
+import Concordium.KonsensusV1.TreeState.Types
+import Concordium.KonsensusV1.Types
+import qualified Data.Map.Strict as Map
+import qualified Data.Sequence as Seq
+
+-- |Generate the 'EpochBakers' for a genesis block.
+genesisEpochBakers ::
+    ( BlockStateQuery m,
+      GSTypes.BlockState m ~ PBS.HashedPersistentBlockState pv,
+      IsConsensusV1 pv,
+      MPV m ~ pv
+    ) =>
+    PBS.HashedPersistentBlockState pv ->
+    m EpochBakers
+genesisEpochBakers genState = do
+    curFullBakers <- getCurrentEpochBakers genState
+    curFinParams <- getCurrentEpochFinalizationCommitteeParameters genState
+    let _currentEpochBakers = computeBakersAndFinalizers curFullBakers curFinParams
+    let _previousEpochBakers = _currentEpochBakers
+    nextFullBakers <- getNextEpochBakers genState
+    nextFinParams <- getNextEpochFinalizationCommitteeParameters genState
+    let _nextEpochBakers = computeBakersAndFinalizers nextFullBakers nextFinParams
+    _nextPayday <- getPaydayEpoch genState
+    return $! EpochBakers{..}
+
+-- |Construct the epoch bakers for a given last finalized block based on the low level tree state
+-- store.
+makeEpochBakers ::
```

Comment:
Instead of this binary search we could also choose to persist `rsLastEpochFinalizationEntry` in the `PersistentRoundStatus` and use the block in one of those QC's and create the bakers directly from there.

Reply:
Perhaps, but this is fine.